x-pack/plugins/task_manager/server/lib/throughput_manager.ts

```diff
@@ -0,0 +1,147 @@
+import { Logger } from '../types';
+import { SavedObjectsErrorHelpers } from '../../../../../src/core/server';
+
+const ADJUST_THROUGHPUT_INTERVAL = 10 * 1000;
+
+// When errors occur, reduce maxWorkers by MAX_WORKERS_DECREASE_PERCENTAGE
+// When errors no longer occur, start increasing maxWorkers by MAX_WORKERS_INCREASE_PERCENTAGE
+// until starting value is reached
+const MAX_WORKERS_DECREASE_PERCENTAGE = 0.8;
+const MAX_WORKERS_INCREASE_PERCENTAGE = 1.05;
+
+// When errors occur, increase pollInterval by POLL_INTERVAL_INCREASE_PERCENTAGE
+// When errors no longer occur, start decreasing pollInterval by POLL_INTERVAL_DECREASE_PERCENTAGE
+// until starting value is reached
+const POLL_INTERVAL_DECREASE_PERCENTAGE = 0.95;
+const POLL_INTERVAL_INCREASE_PERCENTAGE = 1.2;
+
+interface DynamicConfiguration<T> {
+  startingValue: T;
+  currentValue: T;
+  observable$: BehaviorSubject<T>;
+}
+
+interface ThroughputManagerOpts {
+  maxWorkers$: BehaviorSubject<number>;
+  pollInterval$: BehaviorSubject<number>;
+  startingMaxWorkers: number;
+  startingPollInterval: number;
+  errors$: Observable<Error>;
+  logger: Logger;
+}
+
+export class ThroughputManager {
+  private isStarted: boolean = false;
+  private errorsSubscription?: Subscription;
+  private errorCountSinceLastInterval: number = 0;
+  private throughputCheckIntervalId?: NodeJS.Timeout;
+  private readonly logger: Logger;
+  private readonly maxWorkers: DynamicConfiguration<number>;
+  private readonly pollInterval: DynamicConfiguration<number>;
+  private readonly errors$: Observable<Error>;
+
+  constructor(opts: ThroughputManagerOpts) {
+    this.logger = opts.logger;
+    this.errors$ = opts.errors$;
+    this.maxWorkers = {
+      startingValue: opts.startingMaxWorkers,
+      currentValue: opts.startingMaxWorkers,
+      observable$: opts.maxWorkers$,
+    };
+    this.pollInterval = {
+      startingValue: opts.startingPollInterval,
+      currentValue: opts.startingPollInterval,
+      observable$: opts.pollInterval$,
+    };
+  }
+
+  public start() {
+    if (!this.isStarted) {
+      this.errorsSubscription = this.errors$.subscribe(this.storeErrorHandler.bind(this));
+      this.throughputCheckIntervalId = setInterval(
+        this.adjustThroughput.bind(this),
+        ADJUST_THROUGHPUT_INTERVAL
+      );
+      this.isStarted = true;
+    }
+  }
+
+  public stop() {
+    this.errorsSubscription?.unsubscribe();
+    if (this.throughputCheckIntervalId) {
+      clearInterval(this.throughputCheckIntervalId);
+    }
+    delete this.errorsSubscription;
+    delete this.throughputCheckIntervalId;
+    // Reset observable values to original values
+    this.updateConfiguration(this.maxWorkers.startingValue, this.pollInterval.startingValue);
+    this.isStarted = false;
+  }
+
+  private storeErrorHandler(e: Error) {
+    if (SavedObjectsErrorHelpers.isTooManyRequestsError(e)) {
+      this.errorCountSinceLastInterval++;
+    }
+  }
```

Comment:
I love the concept (really do!), but struggling a bit with the flow of this component of the system.
We're ending up with a mix of observable and imperative that feels weird to me.

For example, the `errors$` could be pipes through a stream that filters by `isTooManyRequestsError` and then the count at the end would be  the `errorCountSinceLastInterval` value that gets reset based on an interval stream (instead of the separate `throughputCheckIntervalId` interval).

That kind of model makes the whole thing a single stream that can be subscribed to and wouldn't require you to manage subscriptions and clear intervals. That would also make integrating a `start`/`stop` mechanism easier later when we need to respond to losing ES connectivity and stuff like that. 🤔 

I generally find those kinds of compositions easier to follow in a reactive system, rather than the mix we have here... this isn't a blocker for me, as lots of TM is already imperative and so it isn't out of place, but I won't lie - I find this harder to follow 🤷‍♂️ 

If you'd like to give making this more reactive a shot, I'm very happy to pair with you or guide you through it. :)